### PR TITLE
Use `objc2` and its framework crates

### PR DESCRIPTION
**Pull request by madsmtm**
This catches a few extra error cases related to paths, as well as making it easier to ensure that memory management rules are upheld.

Concretely, it fixes a leak in the passing of the `NSArray` to `writeObjects`.

### CODE REVIEW

File: src/osx_clipboard.rs
Comment (chrisduerr) [inline src/osx_clipboard.rs:39]: Does `generalPasteboard` panic if it's null? This sounds like we're just not calling it because it could return null, but so does msg_send_id.

Comment (madsmtm) [reply src/osx_clipboard.rs:39]: `+[NSPasteboard generalPasteboard]` is specified to never return NULL, but since Objective-C is a lot more loose in that regard than Rust, it may still be NULL in reality (e.g. in case of an allocation error). To ensure soundness in the case of a mismatch like this, `objc2-app-kit` will automatically `unwrap` the pointer for you.

If you make `msg_send_id!` return an `Option` instead, it will not unwrap the value automatically, and we get to do so ourselves. This is somewhat documented [here](https://docs.rs/objc2/0.5.1/objc2/macro.msg_send_id.html#specification).

I doubt this error case is common, I mostly did it to keep roughly the same behaviour as the original code.

**Comment by madsmtm** [comment]
> What happened to objc, is it not communicative? Seems like you're the maintainer of the objc2 crates, is that correct?

Apologies, should've given a more detailed explanation, assumed you knew it already. I'm the maintainer of [`objc2`](https://github.com/madsmtm/objc2), which we've used for a while in Winit. The fork is a few years old now, the reasoning behind it can be found in https://github.com/SSheldon/rust-objc/issues/101. Nowadays, `objc2` contains a lot more features, including the `msg_send_id!` macro described above that ensures we follow memory management rules correctly.

Did that give a bit more insight? Feel free to ask about anything else that's unclear!

File: src/osx_clipboard.rs
Comment (chrisduerr) [reply src/osx_clipboard.rs:39]: Considering crates shouldn't crash, this is probably the best choice.

**Comment by nixpulvis** [comment]
Doing the same things as winit makes a lot of sense since alacirtty is a consumer of both this crate and winit.

**Comment by chrisduerr** [comment]
> the reasoning behind it can be found in SSheldon/rust-objc#101

Sounds like you just got ignored lol.

**Review by chrisduerr — APPROVED** [review]
@madsmtm I'd assume you've tested this against some client? Or used the examples to confirm it actually works?

I have no way to test macOS myself, so I'll hold off merging until I can get a confirmation on this.

**Comment by nixpulvis** [comment]
The `hello_world` example seems to work. I'd be happy to test more thoroughly, but I'm not sure exactly how to trigger the NULLs this PR is primarily concerned with.

Happy to add some tests in the process if given guidance.

**Comment by chrisduerr** [comment]
> I'm not sure exactly how to trigger the NULLs this PR is primarily concerned with.

Shouldn't really be possible, so doesn't need testing. Thanks for the help.

**Pull request merged**
